torch/csrc/dynamo/eval_frame.c

```diff
@@ -438,10 +582,14 @@ static PyObject* lookup(CacheEntry* e, THP_EVAL_API_FRAME_OBJECT *frame, CacheEn
     // If the hit cache entry is not the head of the linked list,
     // move it to the head
     if (prev != NULL) {
-        CacheEntry* extra = get_cache_entry(frame->f_code);
+        ExtraState* extra = get_extra_state(frame->f_code);
+        CacheEntry* old_cache_entry = extract_cache_entry(extra);
+
         prev->next = e->next;
-        e->next = extra;
-        set_cache_entry(frame->f_code, e);
+        e->next = old_cache_entry;
```

Comment:
This is still a violation of borrowing semantics. You're not allowed to steal something you got from a borrow (`extract_cache_entry`). Probably good enough to just say `extra->cache_entry` instead.